README.md
pushwechat![travis-ci](https://secure.travis-ci.org/dead-horse/weixin-push.png)
====== 

 通过模拟后台登录的形式，进行消息的发送和获取。   

## 用法  

```js

var Pusher = require('pwechat');
var pusher = Pusher.create('youremail', 'yourpassword');

pusher.on('PWwchatError', function (err) {
  console.log(err); //无法登录后台
});

/**
 * 给单个用户发送微信消息
 * @param {String} fakeId 用户fakeId
 * @param {String} content 发送内容
 */
pusher.singleSend('12345', 'test content', function (err, data) {
  // 发送成功的响应data.should.eql({ret: 0, msg: 'ok'});
});

//除了可以发送文字外，还支持发送其他的微信资源（在微信素材管理中添加）。  
//type为发送的不同资源类型，不同的类型需要的字段也不同
//发送前需要把要发送的资源上传到微信，并找到对应的`fid`等字段

pusher.singleSend('12345', {
  type: 2, //图片
  fid: 1000002 //图片的资源id
}, function (err, data) {
  // 发送成功的响应data.should.eql({ret: 0, msg: 'ok'});
});

pusher.singleSend('12345', {
  type: 10, //图文
  fid: 1000003,
  fileid: 1000004, 
  appmsgid: 1000003
}, function (err, data) {
  // 发送成功的响应data.should.eql({ret: 0, msg: 'ok'});
});

/**
 * 获取包含关键字的消息
 * @param {String} keyword 消息中包含的关键字
 * @param {Number} count 获取条数
 * @param {Number} fromMsgId 从这个msgId开始往前查找
 */
pusher.getMessage('@help', 10, 1000, function (err, data) {
  // 获取成功的响应，data会是一个数组
});

/**
 * 获取公众账号的粉丝
 * 响应： 
  [{ 
    fakeId: '98106560',
    nickName: 'nick1',
    remarkName: '',
    groupId: '0' 
  }, { 
    fakeId: '3297485',
    nickName: 'nick2',
    remarkName: '',
    groupId: '0' 
  }]
 */
pusher.getUsers(function (err, data) {});

/**
 * 获取单个用户详情, 包括所有分组
 * @param {String}   fakeId   用户的fakeId
 * @param {Function} callback
 * 响应:
{ "FakeId" : "12345",
  "NickName" : "Jane Doe",
  "ReMarkName" : "",
  "Username" : "wxid_12345",
  "Signature": "",
  "Country" : "中国",
  "Province" : "河南",
  "City" : "安阳",
  "Sex" : "1",
  "GroupID" : "0",
  "Groups"  : [
    { "GroupId": "0", "GroupName": "未分组" },
    { "GroupId": "1", "GroupName": "黑名单" },
    { "GroupId": "2", "GroupName": "星标组" }
  ]
}
 */
pusher.getUserDetail('12345', function (err, data) {});

/**
 * 修改用户分组, 每个用户只能分一组
 * @param  {[String]}   fakeId 用户的fakeId
 * @param  {[String]}   userGroupId 分组Id
 * @param  {Function} callback
 * 响应:
{ "ret":"0",
  "result": [
    {"fakeId":"12345","ret":"0"}
  ]
}
 */
pusher.modifyUserGroup('12345', '100', function (err, data) {});


/**
 * 添加分组
 * @param {[String]}   groupName 分组名称
 * @param {Function} callback 
 * 响应:
{ "ErrCode":"",
  "ErrMsg":"",
  "GroupId":"103",
  "GroupName":"apple",
  "MemberCnt":"0"
}
 */
pusher.addGroup('apple', function (err, data) {});

/**
 * 重命名分组
 * @param  {[String]}   groupId   分组ID
 * @param  {[String]}   groupName 新的分组名称
 * @param  {Function}   callback
 * 响应:
{ "ErrCode":"",
  "ErrMsg":"",
  "GroupId":"100",
  "GroupName":"hello",
  "MemberCnt":"0"
}
 */
pusher.renameGroup('100', 'orange', function (err, data) {});


/**
 * 删除分组
 * @param  {[String]}   groupId   分组ID
 * @param  {Function}   callback
 * 响应:
{ "ErrCode":"",
  "ErrMsg":"",
  "GroupId":"100",
  "GroupName":"",
  "MemberCnt":"0"
}
 */
pusher.deleteGroup('100', function (err, data) {});


/**
 * 获取公众账号的上载媒体列表
 * @param {String} type 媒体类别 2为图片, 3为语音, 4为视频, 10为图文
 * @param {Function} callback
 * 响应:
[
{ id: '10000000',
  fileName: '2013082810294238334.jpg',
  type: '2',
  size: '30.6 K',
  dateTime: 1378618280,
  length: '0' }
{ id: '10000004',
  fileName: '2013082608430532299.gif',
  type: '2',
  size: '3.0 K',
  dateTime: 1378623899,
  length: '0' }
{ id: '10000006',
  fileName: '0.jpg',
  type: '2',
  size: '55.3 K',
  dateTime: 1378780173,
  length: '0' }
]
 */
pusher.getMedia('2', function (err, data) {});
```

## 安装  

```
npm install pwechat
```  

## 贡献者们
$ git summary 

```
 project  : weixin-push
 repo age : 7 months ago
 commits  : 40
 active   : 14 days
 files    : 13
 authors  : 
    22  dead-horse              55.0%
     9  不四                  22.5%
     5  dead_horse              12.5%
     2  Angela Zou              5.0%
     1  Xiayu                   2.5%
     1  hsinglin                2.5%
```

## Lincense  
(The MIT License)

Copyright (c) 2012 dead-horse and other contributors

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.



History.md

0.1.1 / 2013-11-01 
==================

  * fix singleSend referer

0.1.0 / 2013-10-03 
==================

  * Merge pull request #11 from angelazou/master
  * Add Groups

0.0.9 / 2013-10-03 
==================

  * add readme
  * fix style
  * Merge pull request #10 from angelazou/master
  * Add getMedia method

0.0.8 / 2013-09-08 
==================

  * fixed #9 get user error

0.0.7 / 2013-08-18 
==================

  * 添加登录请求的header  

0.0.6 / 2013-07-03 
==================

  * 增加查询当前用户fakeId和昵称的json

0.0.5 / 2013-07-03 
==================

  * support more type, fixed #3

0.0.4 / 2013-05-03 
==================

  * fix login token 

0.0.1  
==================  

  * Initial release 
  * add getMessage
  * add singleSend 


test/proxy.test.js
/*!
 * pushwechat - lib/proxy.js 
 * Author: dead-horse <dead_horse@qq.com>
 */


/**
 * Module dependencies.
 */
var proxy = require('../').proxy;
var mm = require('mm');
var should = require('should');
var urllib = require('urllib');

describe('lib/proxy.js', function () {
  describe('#login', function () {
    afterEach(mm.restore);
    it('should response response error', function (done) {
      mm.data(urllib, 'request', '{a: 1}');
      proxy.login('test@gmail.com', '123', function (err, data) {
        err.message.should.equal('response error');
        should.not.exist(data);
        done();
      });
    });

    it('should response login error', function (done) {
      mm(urllib, 'request', function (url, opts, callback) {
        process.nextTick(function () {
          callback(new Error('mock error'), '{"a": "b"}');
        });
      });
      proxy.login('test@gmail.com', '123', function (err, data) {
        err.message.should.equal('mock error');
        should.not.exist(data);
        done();
      });      
    });

    it('should response ok', function (done) {
      mm(urllib, 'request', function (url, opts, callback) {
        process.nextTick(function () {
          callback(null, '{"ErrCode": 0}', {statusCode: 200, headers: {'set-cookie': ['a=b;'] }});
        }); 
      });
      proxy.login('test@gmail.com', '123', function (err, data) {
        should.not.exist(err);
        data.should.equal('a=b;');
        done();
      });
    });
  });

  describe('#singleSend', function () {
    afterEach(mm.restore);

    it('should reponse err', function (done) {
      mm.error(urllib, 'request', 'mock error');
      proxy.singleSend('123', 'test', 'cookie', 'token', function (err) {
        err.message.should.equal('mock error');
        done();
      });
    });

    it('should response error by statusCode', function (done) {
      mm(urllib, 'request', function (url, opts, callback) {
        process.nextTick(function () {
          callback(null, 'test', {statusCode: 302});
        });
      });
      proxy.singleSend('123', 'test', 'cookie', 'token', function (err) {
        err.message.should.equal('singleSend Error! status code:302');
        done();
      });
    });

    it('should response parse error', function (done) {
      mm(urllib, 'request', function (url, opts, callback) {
        process.nextTick(function () {
          callback(null, 'test', {statusCode: 200});
        });
      });
      proxy.singleSend('123', 'test', 'cookie', 'token', function (err) {
        err.message.should.equal('singleSend Error! Unexpected token e');
        done();
      });      
    });

    it('should ok', function (done) {
      mm(urllib, 'request', function (url, opts, callback) {
        process.nextTick(function () {
          callback(null, '{"ret":0, "msg":"ok"}', {statusCode: 200});
        });
      });
      proxy.singleSend('123', 'test', 'cookie', 'token', function (err, data) {
        should.not.exist(err);
        data.should.eql({ret: 0, msg: 'ok'});
        done();
      });      
    });
  });

  describe('#refresh', function () {
    afterEach(mm.restore);
    
    it('should reponse err', function (done) {
      mm.error(urllib, 'request', 'mock error');
      proxy.refresh('cookie', function (err) {
        err.message.should.equal('mock error');
        done();
      });
    });

    it('should response error by statusCode', function (done) {
      mm(urllib, 'request', function (url, opts, callback) {
        process.nextTick(function () {
          callback(null, 'test', {statusCode: 500});
        });
      });
      proxy.refresh('cookie', function (err) {
        err.message.should.equal('refresh Error! status code:500');
        done();
      });
    });

    it('should refresh error by header error', function (done) {
      mm(urllib, 'request', function (url, opts, callback) {
        process.nextTick(function () {
          callback(null, '', {statusCode: 302});
        });
      });
      proxy.refresh('cookie', function (err) {
        err.message.should.equal('Refresh error, can not get token');
        done();
      });
    });

    it('should refresh ok', function (done) {
      mm(urllib, 'request', function (url, opts, callback) {
        process.nextTick(function () {
          callback(null, '', {statusCode: 200, headers: {location: 'token=1234324'}});
        });
      });
      proxy.refresh('cookie', function (err, data) {
        should.not.exist(err);
        data.should.equal('1234324');
        done();
      });
    });
  });

  describe('#getMessage', function () {
    it('should reponse err', function (done) {
      mm.error(urllib, 'request', 'mock error');
      proxy.getMessage('cookie', 'token', '123', 'test', 100, function (err) {
        err.message.should.equal('mock error');
        done();
      });
    });

    it('should response error by statusCode', function (done) {
      mm(urllib, 'request', function (url, opts, callback) {
        process.nextTick(function () {
          callback(null, 'test', {statusCode: 302});
        });
      });
      proxy.getMessage('cookie', 'token', '123', 'test', 100, function (err) {
        err.message.should.equal('getMessage Error! status code:302');
        done();
      });
    });

    it('should response parse error', function (done) {
      mm(urllib, 'request', function (url, opts, callback) {
        process.nextTick(function () {
          callback(null, 'test', {statusCode: 200});
        });
      });
      proxy.getMessage('cookie', 'token', '123', 'test', 100, function (err) {
        err.message.should.equal('getMessage Error! Unexpected token e');
        done();
      });      
    });

    it('should get ok', function (done) {
      mm(urllib, 'request', function (url, opts, callback) {
        process.nextTick(function () {
          callback(null, new Buffer('[]'), {statusCode: 200});
        });
      });
      proxy.getMessage('cookie', 'token', '123', 'test', 100, function (err, data) {
        should.not.exist(err);
        data.should.eql([]);
        done();
      });
    });    
  });
});

test/pushwechat.test.js
/*!
 * pushwechat - lib/proxy.js 
 * Author: dead-horse <dead_horse@qq.com>
 */


/**
 * Module dependencies.
 */
var proxy = require('../').proxy;
var mm = require('mm');
var should = require('should');
var Pusher = require('../');

var pusher;
var loginBack;
describe('lib/pushwechat.js', function () {
  before(function (done) {
    pusher = Pusher.create('testname', 'testpwd', 'token');
    loginBack = proxy.login;
    proxy.login = function (name, pwd, callback) {
      process.nextTick(function () {
        return callback(null, 'mock cookie');
      });
    }
    pusher.once('connect', done);
  });
  after(function () {
    proxy.login = loginBack;
  });

  describe('#singleSend', function () {
    afterEach(mm.restore);
    
    it('should error when proxy error', function (done) {
      mm.error(proxy, 'singleSend', 'mock error');
      pusher.singleSend('fakeId', 'content', function (err, data) {
        err.message.should.equal('mock error');
        should.not.exist(pusher.cookie);
        pusher.cookie = 'mock cookie';
        done();
      });
    });
    
    it('should ok when has cookie', function (done) {
      mm.data(proxy, 'singleSend', {ret: 0, msg: 'ok'});
      pusher.singleSend('fakeId', 'content', function (err, data) {
        should.not.exist(err);
        data.should.eql({ret: 0, msg: 'ok'});
        done();
      });
    });

    it('should ok when no cookie', function (done) {
      mm.data(proxy, 'singleSend', {ret: 0, msg: 'ok'});
      pusher.cookie = null;
      pusher.singleSend('fakeId', 'content', function (err, data) {
        should.not.exist(err);
        data.should.eql({ret: 0, msg: 'ok'});
        pusher.cookie.should.equal('mock cookie');
        done();
      });
    });
  });

  describe('#getMessage', function () {
    afterEach(mm.restore);

    it('should error when proxy error', function (done) {
      mm.error(proxy, 'getMessage', 'mock error');
      pusher.getMessage('fakeId', 'content', 100, function (err, data) {
        err.message.should.equal('mock error');
        should.not.exist(pusher.cookie);
        pusher.cookie = 'mock cookie';
        done();
      });
    });

    it('should ok when has cookie', function (done) {
      mm.data(proxy, 'getMessage', [{}]);
      pusher.getMessage('fakeId', 'content', 100, function (err, data) {
        should.not.exist(err);
        data.should.eql([{}]);
        done();
      });
    });

    it('should error when no cookie and login error', function (done) {
      pusher.cookie = null;
      mm.error(proxy, 'login', 'mock error');
      pusher.getMessage('fakeId', 'content', 100, function (err, data) {
        err.message.should.equal('mock error');
        should.not.exist(pusher.cookie);
        pusher.cookie = 'mock cookie';
        done();
      });
    });

    it('should ok when no cookie', function (done) {
      mm.data(proxy, 'getMessage', [{}]);
      pusher.cookie = null;
      pusher.getMessage('fakeId', 'content', 100, function (err, data) {
        should.not.exist(err);
        data.should.eql([{}]);
        pusher.cookie.should.equal('mock cookie');
        done();
      });
    });
  });

  describe('#_refresh', function () {
    afterEach(mm.restore);
    it('should refresh ok', function (done) {
      pusher._refresh();
      pusher.once('connect', done);
    });

    it('should refresh error', function (done) {
      mm.error(proxy, 'refresh', 'mock error');
      mm.error(proxy, 'login', 'mock error');
      pusher.once('PWechatError', function (err) {
        err.message.should.equal('mock error');
        should.not.exist(pusher.cookie);
        done();
      });
      pusher._refresh();
    })
  });
});

lib/translate.js
/*!
 * pushwechat - lib/proxy.js 
 * Author: dead-horse <dead_horse@qq.com>
 */


/**
 * Module dependencies.
 */

exports.loginErrorMap = function (code) {
  var t = code + "";
  var n;
  switch (t) {
  case "-1":
    n = "系统错误。";
    break;
  case "-2":
    n = "帐号或密码错误。";
    break;
  case "-3":
    n = "密码错误。";
    break;
  case "-4":
    n = "不存在该帐户。";
    break;
  case "-5":
    n = "访问受限。";
    break;
  case "-6":
    n = "需要输入验证码";
    break;
  case "-7":
    n = "此帐号已绑定私人微信号，不可用于公众平台登录。";
    break;
  case "-8":
    n = "邮箱已存在。";
    break;
  case "-32":
    n = "验证码输入错误";
    break;
  case "-200":
    n = "因频繁提交虚假资料，该帐号被拒绝登录。";
    break;
  case "-94":
    n = "请使用邮箱登陆。";
    break;
  case "10":
    n = "该公众会议号已经过期，无法再登录使用。";
    break;
  default:
    n = "未知的返回。";
  }
  return n;
};


lib/proxy.js
/*!
 * pushwechat - lib/proxy.js 
 * Author: dead-horse <dead_horse@qq.com>
 */


/**
 * Module dependencies.
 */
var urllib = require('urllib');
var cheerio = require('cheerio');
var utility = require('utility');
var translate = require('./translate');

var URLS = {
  login: 'http://mp.weixin.qq.com/cgi-bin/login?lang=zh_CN',
  singleSend: 'http://mp.weixin.qq.com/cgi-bin/singlesend?t=ajax-response&lang=zh_CN',
  refresh: 'http://mp.weixin.qq.com/cgi-bin/indexpage',
  getMessage: 'http://mp.weixin.qq.com/cgi-bin/getmessage?t=ajax-message',
  getUsers: 'https://mp.weixin.qq.com/cgi-bin/contactmanage',
  getUserDetail: 'https://mp.weixin.qq.com/cgi-bin/getcontactinfo',
  modifyUserGroup: 'https://mp.weixin.qq.com/cgi-bin/modifycontacts',
  modifyGroup: 'http://mp.weixin.qq.com/cgi-bin/modifygroup',
  getMedia: 'https://mp.weixin.qq.com/cgi-bin/filemanagepage'
};

var LANG = 'zh_CN';
var okStatus = [65201, 65202, 0];
var USER_AGENT = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_8_2) AppleWebKit/537.11 (KHTML, like Gecko) Chrome/23.0.1271.95 Safari/537.11';
var POST_HEADER = 'application/x-www-form-urlencoded; charset=UTF-8';

exports.login = function (username, pwd, callback) {
  var data = {
    username: username,
    pwd: utility.md5(pwd.substr(0, 16)).toLowerCase(),
    imgcode: '',
    f: 'json'
  };
  var headers = {
    'Content-Type': POST_HEADER,
    Host: 'mp.weixin.qq.com',
    Origin: 'https://mp.weixin.qq.com',
    Referer: 'https://mp.weixin.qq.com/',
    'User-Agent' : USER_AGENT
  };  
  urllib.request(URLS.login, {
    type: 'POST',
    data: data,
    headers: headers
  }, function (err, data, res) {
    try {
      data = JSON.parse(data);    
    } catch (err) {
      return callback(new Error('response error'));
    }
    if (err || (res.statusCode !== 200 && res.statusCode !== 302) || okStatus.indexOf(data.ErrCode) < 0) {
      return callback(err || new Error('login error!' + translate.loginErrorMap(data.ErrCode)));
    }
    var cookies = '';
    var token;
    if (data.ErrMsg) {
      var matchs = data.ErrMsg.match(/token=(\d+)/);
      token = matchs ? matchs[1] : '';
    }
    var cookieReg = /^([\w_]+)=([\w_=]+);/;
    res.headers['set-cookie'].forEach(function (cookie) {
      var cmatchs = cookie.match(cookieReg);
      if (cmatchs) {
        cookies += cmatchs[0];
      }
    });
    callback(null, cookies, token);
  });
};

exports.singleSend = function (fakeId, content, cookie, token, callback) {
  var data;
  if (typeof content === 'string') {
    data = {
      type: 1,
      content: content,
      tofakeid: fakeId,
      ajax: 1,
      token: token,
      error: false
    };
  } else {
    data = {
      tofakeid: fakeId,
      ajax: 1,
      token: token,
      error: false
    };
    for (var key in content) {
      data[key] = content[key];
    }
    data.fileid = data.fileid || data.fid;  
  }
  var headers = {
    cookie: cookie,
    'Content-Type': POST_HEADER,
    Referer: 'https://mp.weixin.qq.com/cgi-bin/message?t=message/list'
  };
  urllib.request(URLS.singleSend, {
    type: 'POST',
    headers: headers,
    data: data
  }, function (err, data, res) {
    if (err || res.statusCode !== 200) {
      return callback(err || new Error('singleSend Error! status code:' + res.statusCode));
    }
    try {
      data = JSON.parse(data);
    } catch (err) {
      return callback(new Error('singleSend Error! ' + err.message));
    }
    callback(null, data);
  });
};

exports.refresh = function (cookie, callback) {
  urllib.request(URLS.refresh, {headers: {cookie: cookie}}, function (err, data, res) {
    if (err || (res.statusCode !== 302 && res.statusCode !== 200)) {
      return callback(err || new Error('refresh Error! status code:' + res.statusCode));
    }
    var token;
    if (res.headers && res.headers.location) {
      var matchs = res.headers.location.match(/token=(\d+)/);
      token = matchs ? matchs[1] : '';
    }
    return callback(token ? null : new Error('Refresh error, can not get token'), token);
  });
};

exports.getMessage = function (cookie, token, keyword, count, fromMsgId, callback) {
  var data = {
    keyword: keyword,
    count: count,
    frommsgid: fromMsgId,
    token: token,
    ajax: 1
  };
  var headers = {
    'User-Agent': USER_AGENT,
    Referer: 'https://mp.weixin.qq.com/cgi-bin/getmessage?token=1811982400&t=wxm-message&lang=zh_CN&count=10&keyword=%E6%B5%8B%E8%AF%95',
    Cookie: cookie
  };
  urllib.request(URLS.getMessage, {
    type: 'POST',
    headers: headers,
    data: data
  }, function (err, data, res) {
    if (err || res.statusCode !== 200) {
      return callback(err || new Error('getMessage Error! status code:' + res.statusCode));
    }
    try {
      data = JSON.parse(data);
    } catch (err) {
      return callback(new Error('getMessage Error! ' + err.message));
    }
    callback(null, data);
  });
};

exports.getUsers = function (cookie, token, callback) {
  var data = {
    t:"wxm-friend",
    token:token,
    lang:LANG,
    pagesize:100000,
    pageidx:0,
    type:0,
    groupid:0
  }
  var headers = {
    'User-Agent': USER_AGENT,
    Referer: 'https://mp.weixin.qq.com/cgi-bin/message?t=message/list&count=20&day=7&token=337134409&lang=zh_CN',
    Cookie: cookie
  };
  urllib.request(URLS.getUsers, {
    type: 'GET',
    headers: headers,
    data: data
  }, function (err, data, res) {
    if (err || res.statusCode !== 200) {
      return callback(err || new Error('getUsers Error! status code:' + res.statusCode));
    }
    var friendListReg = /friendsList\s*:\s*\(\{"contacts"\s*:\s*(.*?)\}\)\.contacts/;
    var friendList = data.toString().match(friendListReg);
    if (!friendList) {
      return callback(new Error('Get friend list return error, wechat change format'));
    }
    try {
      friendList = JSON.parse(friendList[1]);
    } catch (err) {
      return callback(new Error('Get friend list return error, wechat change format')); 
    }
    callback(null,friendList);
  });
};

exports.getUserDetail = function(cookie, token, callback, fakeId) {
  var data = {
    t: 'ajax-getcontactinfo',
    lang: LANG,
    token: token,
    fakeid: fakeId
  };
  var headers = {
    Cookie: cookie,
    Referer: 'https://mp.weixin.qq.com/cgi-bin/contactmanage?t=user/index&pagesize=10&pageidx=0&type=0&groupid=0&token=2039772792&lang=zh_CN',
    'User-Agent': USER_AGENT,
    'Content-Type': POST_HEADER,
    'X-Requested-With': 'XMLHttpRequest'
  };
  urllib.request(URLS.getUserDetail, {
    type: 'POST',
    headers: headers,
    data: data
  }, function (err, data, res) {
    if (err || res.statusCode !== 200) {
      return callback(err || new Error('getUserDetail Error! status code:' + res.statusCode));
    }

    data = data.toString().replace(/\s{2,}|\t/g, ' ');

    if (!data) {
      return callback(new Error('Get user detail return error. No Content Obtained'));
    }
    try {
      data = JSON.parse( data );
    } catch (err) {
      return callback(new Error('Get user detail return error, wechat change format')); 
    }

    callback(null, data);
  });
};

exports.modifyUserGroup = function(cookie, token, callback, fakeId, userGroup) {
  var data = {
    t: 'ajax-putinto-group',
    lang: LANG,
    token: token,
    action: 'modifycontacts',
    contacttype: userGroup,
    tofakeidlist: fakeId
  };
  var headers = {
    Cookie: cookie,
    Referer: 'https://mp.weixin.qq.com/cgi-bin/contactmanage?t=user/index&pagesize=10&pageidx=0&type=0&groupid=100&token=2039772792&lang=zh_CN',
    'User-Agent': USER_AGENT,
    'Content-Type': POST_HEADER,
    'X-Requested-With': 'XMLHttpRequest'
  };
  urllib.request(URLS.modifyUserGroup, {
    type: 'POST',
    headers: headers,
    data: data
  }, function (err, data, res) {
    if (err || res.statusCode !== 200) {
      return callback(err || new Error('modifyUserGroup Error! status code:' + res.statusCode));
    }

    data = data.toString().replace(/\s{2,}|\t/g, ' ');

    if (!data) {
      return callback(new Error('Modify User Group return error. No Content Obtained'));
    }
    try {
      data = JSON.parse( data );
    } catch (err) {
      return callback(new Error('Modify User Group return error, wechat change format')); 
    }

    callback(null, data);
  });
};

exports.addGroup = function(cookie, token, callback, groupName) {
  var data = {
    t: 'ajax-friend-group',
    lang: LANG,
    name: groupName,
    func: 'add',
    token: token
  };
  var headers = {
    Cookie: cookie,
    Referer: 'https://mp.weixin.qq.com/cgi-bin/contactmanage?t=user/index&pagesize=10&pageidx=0&type=0&groupid=100&token=2039772792&lang=zh_CN',
    'User-Agent': USER_AGENT,
    'Content-Type': POST_HEADER,
    'X-Requested-With': 'XMLHttpRequest'
  };
  urllib.request(URLS.modifyGroup, {
    type: 'POST',
    headers: headers,
    data: data
  }, function (err, data, res) {
    if (err || res.statusCode !== 200) {
      return callback(err || new Error('addGroup Error! status code:' + res.statusCode));
    }

    data = data.toString().replace(/\s{2,}|\t/g, ' ');

    if (!data) {
      return callback(new Error('Add Group return error. No Content Obtained'));
    }
    try {
      data = JSON.parse( data );
    } catch (err) {
      return callback(new Error('Add Group return error, wechat change format')); 
    }

    callback(null, data);
  }); 
};

exports.renameGroup = function(cookie, token, callback, groupId, groupName) {
  var data = {
    t: 'ajax-friend-group',
    id: groupId,
    lang: LANG,
    name: groupName,
    func: 'rename',
    token: token
  };
  var headers = {
    Cookie: cookie,
    Referer: 'https://mp.weixin.qq.com/cgi-bin/contactmanage?t=user/index&pagesize=10&pageidx=0&type=0&groupid=100&token=2039772792&lang=zh_CN',
    'User-Agent': USER_AGENT,
    'Content-Type': POST_HEADER,
    'X-Requested-With': 'XMLHttpRequest'
  };
  urllib.request(URLS.modifyGroup, {
    type: 'POST',
    headers: headers,
    data: data
  }, function (err, data, res) {
    if (err || res.statusCode !== 200) {
      return callback(err || new Error('renameGroup Error! status code:' + res.statusCode));
    }

    data = data.toString().replace(/\s{2,}|\t/g, ' ');

    if (!data) {
      return callback(new Error('Rename Group return error. No Content Obtained'));
    }
    try {
      data = JSON.parse( data );
    } catch (err) {
      return callback(new Error('Rename Group return error, wechat change format')); 
    }

    callback(null, data);
  });
};

exports.deleteGroup = function(cookie, token, callback, groupId) {
  var data = {
    t: 'ajax-friend-group',
    id: groupId,
    lang: LANG,
    func: 'del',
    token: token
  };
  var headers = {
    Cookie: cookie,
    Referer: 'https://mp.weixin.qq.com/cgi-bin/contactmanage?t=user/index&pagesize=10&pageidx=0&type=0&groupid=100&token=2039772792&lang=zh_CN',
    'User-Agent': USER_AGENT,
    'Content-Type': POST_HEADER,
    'X-Requested-With': 'XMLHttpRequest'
  };
  urllib.request(URLS.modifyGroup, {
    type: 'POST',
    headers: headers,
    data: data
  }, function (err, data, res) {
    if (err || res.statusCode !== 200) {
      return callback(err || new Error('deleteGroup Error! status code:' + res.statusCode));
    }

    data = data.toString().replace(/\s{2,}|\t/g, ' ');

    if (!data) {
      return callback(new Error('Delete Group return error. No Content Obtained'));
    }
    try {
      data = JSON.parse( data );
    } catch (err) {
      return callback(new Error('Delete Group return error, wechat change format')); 
    }

    callback(null, data);
  });
};


exports.getMedia = function(cookie, token, callback, type) {
  var data = {
    t: "wxm-file",
    type: type,
    lang: LANG,
    token: token,
    pagesize: 100000,
    pageidx: 0
  };
  //图文有subtype
  if(type === 10) { 
    data.subtype = 3; 
  }
  var headers = {
    Cookie: cookie,
    'User-Agent': USER_AGENT,
  };
  urllib.request(URLS.getMedia, {
    type: 'GET',
    headers: headers,
    data: data
  }, function (err, data, res) {
    if (err || res.statusCode !== 200) {
      return callback(err || new Error('getMedia Error! status code:' + res.statusCode));
    }
    $ = cheerio.load(data.toString());
    var mediaList = $("#json-fileList").html();

    if (!mediaList) {
      return callback(new Error('Get media list return error. No Content Obtained'));
    }
    try {
      //干掉坑爹的tab
      mediaList = JSON.parse( mediaList.replace(/\s{2,}|\t/g, ' ') );
    } catch (err) {
      return callback(new Error('Get media list return error, wechat change format')); 
    }
    callback(null, mediaList);
  });
};